app/src/main/java/agency/five/codebase/android/movieapp/ui/home/HomeViewModel.kt

```diff
@@ -0,0 +1,127 @@
+        getUpcomingMovies(homeScreenMapper)
+    }
+
+    private fun getPopularMovies(mapper: HomeScreenMapper) {
+        viewModelScope.launch {
+            movieRepository.popularMovies(MovieCategory.POPULAR_STREAMING).collect {
+                _popularViewState.value = mapper.toHomeMovieCategoryViewState(
+                    movieCategories = popularCategoryLabels,
+                    selectedMovieCategory = MovieCategory.POPULAR_STREAMING,
+                    movies = it
+                )
+            }
+        }
+    }
+
+    private fun getNowPlayingMovies(mapper: HomeScreenMapper) {
+        viewModelScope.launch {
+            movieRepository.nowPlayingMovies(MovieCategory.NOWPLAYING_MOVIES).collect {
+                _nowPlayingViewState.value = mapper.toHomeMovieCategoryViewState(
+                    movieCategories = nowPlayingCategoryLabels,
+                    selectedMovieCategory = MovieCategory.NOWPLAYING_MOVIES,
+                    movies = it
+                )
+            }
+        }
+    }
+
+    private fun getUpcomingMovies(mapper: HomeScreenMapper) {
+        viewModelScope.launch {
+            movieRepository.upcomingMovies(MovieCategory.UPCOMING_TODAY).collect {
+                _upcomingViewState.value = mapper.toHomeMovieCategoryViewState(
+                    movieCategories = upcomingCategoryLabels,
+                    selectedMovieCategory = MovieCategory.UPCOMING_TODAY,
+                    movies = it
+                )
+            }
+        }
+    }
+
+    fun onPopularLabelClick(id: Int) {
+        viewModelScope.launch {
+            movieRepository.popularMovies(MovieCategory.POPULAR_STREAMING).collect {
+                _popularViewState.value = homeScreenMapper.toHomeMovieCategoryViewState(
+                    movieCategories = popularCategoryLabels,
+                    selectedMovieCategory = MovieCategory.getByOrdinal(id)!!,
+                    movies = it
+                )
+            }
+        }
+    }
+
+    fun onNowPlayingLabelClick(id: Int) {
+        viewModelScope.launch {
+            movieRepository.nowPlayingMovies(MovieCategory.NOWPLAYING_MOVIES).collect {
+                _nowPlayingViewState.value = homeScreenMapper.toHomeMovieCategoryViewState(
+                    movieCategories = nowPlayingCategoryLabels,
+                    selectedMovieCategory = MovieCategory.getByOrdinal(id)!!,
+                    movies = it
+                )
+            }
+        }
+    }
+
+    fun onUpcomingLabelClick(id: Int) {
+        viewModelScope.launch {
+            movieRepository.upcomingMovies(MovieCategory.UPCOMING_TODAY).collect {
+                _upcomingViewState.value = homeScreenMapper.toHomeMovieCategoryViewState(
+                    movieCategories = upcomingCategoryLabels,
+                    selectedMovieCategory = MovieCategory.getByOrdinal(id)!!,
+                    movies = it
+                )
+            }
+        }
+    }
```

Comment:
This should just be one function which will then select on which flow it will emit, and then each view state flow will change when this emits. So you will have something like `popularMoviesCategorySelected`, which will check which popular category was selected. The view state flows, like `popularViewState` will then be mapped from popular moves and call movie repository method. 

The structure will be something like this:
```kotlin
    val popularCategoryViewState = popularMoviesCategorySelected
            .flatMapLatest { selectedMovieCategory ->
                movieRepository.popularMovies(selectedMovieCategory)
                    .map { ... }
            }
```